pkg/virtualizers/firecracker/util_linux.go

```diff
@@ -0,0 +1,331 @@
+
+		}
+		io.Copy(w, bytes.NewBuffer(body))
+	}
+}
+
+// OrganiseTapDevices handles http requests to create and delete tap interfaces for firecracker
+func OrganiseTapDevices(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case http.MethodPost:
+		createDevices(w, r)
+	case http.MethodDelete:
+		deleteDevices(w, r)
+	default:
+		http.Error(w, "method not available", http.StatusBadRequest)
+	}
+}
+
+func deleteDevices(w http.ResponseWriter, r *http.Request) {
+	var dd Devices
+	err := json.NewDecoder(r.Body).Decode(&dd)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+
+	for i := 0; i < len(dd.Devices); i++ {
+		err := tenus.DeleteLink(dd.Devices[i])
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+		}
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+// writeCounter counts the number of bytes written to it.
+type writeCounter struct {
+	total      int64 // total size
+	downloaded int64 // downloaded # of bytes transferred
+	onProgress func(downloaded int64, total int64)
+}
+
+// Write implements the io.Writer interface.
+//
+// Always completes and never returns an error.
+func (wc *writeCounter) Write(p []byte) (n int, e error) {
+	n = len(p)
+	wc.downloaded += int64(n)
+	wc.onProgress(wc.downloaded, wc.total)
+	return
+}
+func newWriter(size int64, onProgress func(downloaded, total int64)) io.Writer {
+	return &writeCounter{total: size, onProgress: onProgress}
+}
+
+func (o *operation) fetchLength(file *os.File, client *http.Client, url string, kernel string) (int, error) {
+	// Determinate the file size
+	resp, err := client.Head(url)
+	if err != nil {
+		os.Remove(file.Name())
+		return 0, err
+	}
+	if resp.StatusCode == 404 {
+		os.Remove(file.Name())
+		return 0, fmt.Errorf("Kernel '%s' VMLinux does not exist", kernel)
+	}
+	length, err := strconv.Atoi(resp.Header.Get("content-length"))
+	if err != nil {
+		return 0, err
+	}
+	return length, nil
+}
+func (o *operation) createFileAndGetLength(kernel string, client *http.Client, url string) (*os.File, int, error) {
+	// Create file locally to download
+	file, err := os.Create(filepath.Join(o.firecrackerPath, kernel))
+	if err != nil {
+		return nil, 0, err
+	}
+	defer file.Close()
+
+	length, err := o.fetchLength(file, client, url, kernel)
+	if err != nil {
+		os.Remove(file.Name())
+		return nil, 0, err
+	}
+
+	return file, length, nil
+}
+
+// fetchVMLinux reads the kernel it wants to run and returns the vm linux required to run
+// Will download the vmlinux if it doesn't exist
+func (o *operation) fetchVMLinux(kernel string) (string, error) {
+	o.updateStatus(fmt.Sprintf("Fetching VMLinux searching %s for %s", o.firecrackerPath, kernel))
+	// check if vmlinux is on system at valid path
+	_, err := os.Stat(filepath.Join(o.firecrackerPath, kernel))
+	if err != nil {
+		// file doesn't exist must download from bucket
+		o.updateStatus(fmt.Sprintf("VMLinux for kernel doesn't exist downloading..."))
```

Comment:
messages is a bit weird. Comma missing? "VMLinux for kernel doesn't exist downloading..."